Apple .strings support

magyarosibotond opened the pull request:
Add Apple .strings file support support as proposed in #13 by @BrunoMiguens.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/traduora/traduora?pullRequest=17) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/traduora/traduora?pullRequest=17) before we can accept your contribution.<br/><hr/>**Botond Magyarosi** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/traduora/traduora?pullRequest=17) it.</sub>

anthonynsimon commented:
@magyarosibotond looks good! I see that you already signed the CLA in Google Docs, but our cla-bot seems to be having trouble associating the commits, did you already add the email you used on your commits on GitHub?

Maybe this helps: https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

Other than that, I'll give it a try soon and we can merge it.

anthonynsimon commented on api/src/formatters/strings.spec.ts:
```diff
@@ -0,0 +1,15 @@
+import { loadFixture, simpleFormatFixture } from "./fixtures";
+import { stringsParser, stringsExporter } from "./strings";
+import { async } from "rxjs/internal/scheduler/async";
```
Could you remove this import? The editor probably added it automatically.

magyarosibotond replied:
rebased

anthonynsimon commented on api/src/controllers/import.controller.ts:
```diff
@@ -32,6 +32,8 @@ import { yamlFlatParser } from '../formatters/yaml-flat';
 import AuthorizationService from '../services/authorization.service';
 import { gettextParser } from '../formatters/gettext';
+import { stringsParser } from '../formatters/strings';
+import { from } from 'rxjs';
```
Is the import `rxjs` needed ? :)

magyarosibotond replied:
rebased

magyarosibotond commented:
Hey @anthonynsimon! 

First of all, thanks for open-sourcing this great tool! I'm more than happy to contribute to this project!

Removed the unused imports of rxjs and rebased my commits to link to my GitHub account (git wasn't configured correctly on this machine for GitHub) but it seems like CLAassistant still can't see my agreement.

Any idea?

anthonynsimon commented:
@magyarosibotond thanks for the feedback! Regarding the cla bot, I think you might just have to sign it again but via the bot, I now removed the google docs option since it was a bit flaky :)

magyarosibotond commented:
@anthonynsimon Done! I've also updated the documentation for .strings.

anthonynsimon commented:
@magyarosibotond Looks good!

I tried it locally via the webapp and it worked well with several different files I have. However the tests were failing (unfortunately circleCI seems to not be picking up the PR haha).

If you run the tests via `bin/test.sh` you will see what I mean, it looked like the fixture file is not exactly what the output of the exporter does, I think the fixture should become:

```

"term.one" = "Current Plan: {{ project.plan.name }}";

"term two" = "{VAR_PLURAL, plural, =0 {locales} =1 {locale} other {locales} }";

"TERM_THREE" = "Export format...";

"term:four" = "hello there you\nthis should be in a newline";

```

Notice the newlines before and after too.

Once you've fixed that, just do the two following and you're good to go!

- You can add a new entry under the `next` section in the changelog about this new feature
- Run `yarn fmt` inside the api directory and commit those too

Thanks a lot for contributing!

magyarosibotond commented:
@anthonynsimon Those extra lines might be reserved for comments in Apple's .strings files (comment for translator). Check the documentation:
https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/LoadingResources/Strings/Strings.html

If I'm reading it correctly, the 3rd party doesn't support empty line free output.

Should I run a regex against the output or remove the dependency and add the lines formatting the output?

anthonynsimon commented:
@magyarosibotond ah I see, wouldn't it be fine to just update the fixture to match the output (with the additional newlines)? I think it should be fine, both tests pass with that change.

magyarosibotond commented:
@anthonynsimon As an iOS developer we prefer denser strings files. I'll remove the unnecessary empty lines and re-add when translator comments will be added as a feature.

magyarosibotond commented:
Fixed

anthonynsimon commented:
Thank you!
